File: insights-operator-runtime/src/insights_operator_runtime/file.rs
use log::debug;
use rc_zip_sync::ReadZip;
use std::collections::HashMap;
use std::env::split_paths;
use std::fs::{self, File};
use std::io::BufRead;
use std::io::{self, Write};
use std::os::unix::fs::PermissionsExt;
use std::path::Path;
use std::path::PathBuf;
use std::process::Command;
use std::process::Stdio;

/// Create a directory and return its File.
/// Remove the directory if it exists before creating it.
///
/// The directory is created with 777 permissions so that anybody can write into it.
pub fn create_dir(name: &str) -> io::Result<File> {
    debug!("📂  (re)creating dir {}", name);

    if let Err(e) = fs::remove_dir_all(&name) {
        if e.kind() != io::ErrorKind::NotFound {
            return Err(e);
        }
    }
    fs::create_dir(name)?;
    fs::set_permissions(&name, fs::Permissions::from_mode(0o777))?;

    File::open(&name)
}

/// Read a key=value file and return its content in a HashMap.
///
/// Key and values are separated by '='.
pub fn read_key_value_file(path: &str) -> io::Result<HashMap<String, String>> {
    debug!("📄  Reading key=value file {:?}", path);

    let content = fs::read_to_string(path)?;
    let mut map = HashMap::new();

    for line in content.lines() {
        let parts: Vec<&str> = line.splitn(2, '=').collect();
        if parts.len() == 2 {
            let key = parts[0].trim().to_string();
            let value = parts[1].trim_matches('"').to_string();
            map.insert(key, value);
        }
    }

    Ok(map)
}

/// Read a Jar manifest and return its entries in a HashMap
///
/// Keys and values are separated by ':'.
pub fn read_jar_manifest(manifest_file: &str) -> io::Result<HashMap<String, String>> {
    debug!("📄  Reading jar manifest {:?}", manifest_file);

    let content = fs::read_to_string(manifest_file)?;

    read_jar_manifest_content(&content)
}

/// Read a Jar manifest and return its entries in a HashMap
///
/// Keys and values are separated by ':'.
fn read_jar_manifest_content(manifest_content: &str) -> io::Result<HashMap<String, String>> {
    // /// 	manifestEntries := make(map[string]string)
    // currentKey := ""
    // for _, entry := range strings.Split(strings.ReplaceAll(string(b), "\r\n", "\n"), "\n") {
    // 	key, value, keyFound := strings.Cut(entry, ": ")
    // 	if keyFound {
    // 		manifestEntries[key] = value
    // 		currentKey = key
    // 	} else {
    // 		manifestEntries[currentKey] = manifestEntries[currentKey] + strings.TrimLeft(entry, " ")
    // 	}
    // }
    // return manifestEntries, nil

    let mut map = HashMap::new();

    let mut current_key = "";
    for line in manifest_content.lines() {
        let parts: Vec<&str> = line.splitn(2, ':').collect();
        if parts.len() == 2 {
            let key = parts[0].trim();
            let value = parts[1].trim().trim_matches('"').to_string();
            map.insert(key.to_string(), value);
            current_key = key;
        } else if parts.len() == 1 {
            let value = parts[0].trim().trim_matches('"').to_string();
            let existing_value = map.get(current_key);
            map.insert(
                current_key.to_string(),
                format!("{}{}", existing_value.unwrap(), value),
            );
        }
    }

    Ok(map)
}

pub fn get_jar_manifest(jar_path: &str) -> io::Result<HashMap<String, String>> {
    debug!("📄  Reading manifest for jar {:?}", jar_path);
    eprintln!("📄  Reading manifest for jar {:?}", jar_path);

    if let Ok(jar_file) = File::open(jar_path) {
        eprintln!("got jar file {:#?}", jar_file);
        if let Ok(archive) = jar_file.read_zip() {
            eprintln!("got archive");
            if let Some(manifest) = archive.by_name("META-INF/MANIFEST.MF") {
                let lines = io::BufReader::new(manifest.reader()).lines();
                let content: String = lines
                    .map(|line| line.map(|l| l + "\n"))
                    .collect::<Result<_, _>>()?;
                println!("content: {:#?}", content);
                return read_jar_manifest_content(&content);
            }
        }
    }

    Ok(HashMap::new())
}

pub fn jar_contains_class(jar_path: &PathBuf, class_name: &str) -> bool {
    let class_file = class_name.replace(".", "/") + ".class";

    debug!("📄  List classes in the jar {:?}", jar_path);
    eprintln!("📄  Reading manifest for jar {:?}", jar_path);
    if let Ok(jar_file) = File::open(jar_path) {
        if let Ok(archive) = jar_file.read_zip() {
            match archive.entries().find(|file| file.name == class_file) {
                Some(_) => {
                    return true;
                }
                None => {}
            }
        }
    }

    false
}

/// Write fingerprints entries to a file in the `out` directory.
///
/// The file is named `${prefix}-fingerprints.txt`.
pub fn write_fingerprint(
    out: &Path,
    prefix: &str,
    entries: &HashMap<String, String>,
) -> io::Result<()> {
    if entries.len() == 0 || prefix.len() == 0 {
        return Ok(());
    }

    let binding = Path::new(&out).join(format!("{}-fingerprints.txt", prefix));
    let fingerprint_file = binding.as_path();
    let mut file = File::create(fingerprint_file)?;

    let mut keys: Vec<_> = entries.keys().collect();
    keys.sort();

    for key in keys {
        if let Some(value) = entries.get(key) {
            writeln!(&mut file, "{}={}", key, value)?;
        }
    }

    debug!("📄  Wrote fingerprints file {:?}", fingerprint_file);

    Ok(())
}

pub fn get_executable_version_output(exec: &str) -> io::Result<String> {
    let output = Command::new(&exec)
        .arg("--version")
        .stdout(Stdio::piped())
        .output();

    match output {
        Ok(output) => {
            if output.status.success() {
                if let Ok(output_str) = String::from_utf8(output.stdout) {
                    Ok(output_str.trim().to_string())
                } else {
                    let custom_error =
                        io::Error::new(io::ErrorKind::Other, "No output from command");
                    Err(custom_error)
                }
            } else {
                let custom_error = io::Error::new(io::ErrorKind::Other, "command did not succeed");
                Err(custom_error)
            }
        }
        Err(err) => Err(err),
    }
}

pub fn recursively_resolve_symlink(path: PathBuf) -> io::Result<PathBuf> {
    let mut current_path = path;
    loop {
        let metadata = fs::symlink_metadata(&current_path)?;
        if metadata.file_type().is_symlink() {
            current_path = fs::read_link(&current_path)?;
        } else {
            return Ok(current_path);
        }
    }
}

pub fn find_executable_in_path(executable_name: &str, path_var: &str) -> Option<PathBuf> {
    for path in split_paths(&path_var) {
        let exe_path = path.join(executable_name);
        if exe_path.is_file() {
            return Some(exe_path);
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashMap;

    use crate::file::create_dir;

    #[test]
    fn it_write_fingerprint() {
        let out = String::from("target/tmp-test");
        create_dir(&out).expect("Create temporary directory");

        let mut entries = HashMap::new();
        entries.insert(String::from("foo"), String::from("value1"));
        entries.insert(String::from("bar"), String::from("value2"));

        let dir_path = Path::new(&out);
        write_fingerprint(dir_path, "test", &entries).expect("wrote fingerprints");

        let content = fs::read_to_string(dir_path.join("test-fingerprints.txt"))
            .expect("Read content from fingerprint file");
        // map is stored in natural order
        assert_eq!(
            content,
            "\
bar=value2
foo=value1
"
            .to_string()
        );

        let _ = fs::remove_dir_all(dir_path);
    }
}

File: fingerprints/pkg/utils/file.go
package utils

import (
	"archive/zip"
	"bufio"
	"bytes"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
)

func WriteEntries(outputDir string, fileName string, entries map[string]string) {
	file, err := os.Create(filepath.Join(outputDir, fileName))
	if err != nil {
		panic(err)
	}
	defer file.Close()

	keys := make([]string, 0, len(entries))
	for k := range entries {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		file.WriteString(k + "=" + entries[k] + "\n")
	}
}

// / Read a key=value file and return its content in a map.
// /
// / Key and values are separated by '='.
func ReadPropertiesFile(filePath string) (map[string]string, bool) {
	_, err := os.Stat(filePath)
	if os.IsNotExist(err) {
		// File does not exist
		return nil, false
	}

	file, err := os.Open(filePath)
	if err != nil {
		return nil, false
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	properties := make(map[string]string)

	for scanner.Scan() {
		line := scanner.Text()
		if strings.HasPrefix(line, "#") || strings.TrimSpace(line) == "" {
			continue
		}

		// Split the line into key and value
		parts := strings.SplitN(line, "=", 2)
		if len(parts) == 2 {
			key := strings.Trim(strings.TrimSpace(parts[0]), "\"")
			value := strings.Trim(strings.TrimSpace(parts[1]), "\"")
			properties[key] = value
		}
	}

	// Check for scanner errors
	if err := scanner.Err(); err != nil {
		return nil, false
	}

	return properties, true
}

func GetExecutableVersionOutput(executable string) (string, error) {
	out, err := exec.Command(executable, "--version").Output()
	if err != nil {
		return "", err
	}
	return string(out), nil
}

// FindExecutableInPath returns the directory in the $PATH env var that contains the given executable
func FindExecutableInPath(executable string, pathEnvVar string) (string, error) {
	paths := strings.Split(pathEnvVar, string(os.PathListSeparator))
	for _, dir := range paths {
		fullPath := filepath.Join(dir, executable)
		if _, err := os.Stat(fullPath); err == nil {
			return fullPath, nil
		}
	}
	return "", fmt.Errorf("executable %s not found in PATH", executable)
}

func GetJarManifest(jarPath string) (map[string]string, error) {
	r, err := zip.OpenReader(jarPath)
	if err != nil {
		return nil, fmt.Errorf("failed to open JAR file: %w", err)
	}
	defer r.Close()

	for _, file := range r.File {
		if file.Name == "META-INF/MANIFEST.MF" {
			manifest, err := file.Open()
			if err != nil {
				return nil, fmt.Errorf("failed to open manifest file: %w", err)
			}
			defer manifest.Close()

			var buffer bytes.Buffer
			if _, err := io.Copy(&buffer, manifest); err != nil {
				return nil, fmt.Errorf("failed to read manifest file: %w", err)
			}

			manifestContent := buffer.String()

			manifestEntries := make(map[string]string)
			currentKey := ""
			for _, entry := range strings.Split(strings.ReplaceAll(manifestContent, "\r\n", "\n"), "\n") {
				key, value, keyFound := strings.Cut(entry, ": ")
				if keyFound {
					manifestEntries[key] = value
					currentKey = key
				} else {
					manifestEntries[currentKey] = manifestEntries[currentKey] + strings.TrimLeft(entry, " ")
				}
			}
			return manifestEntries, nil
		}
	}

	return nil, fmt.Errorf("manifest file not found in jar %s", jarPath)
}

func JarFileContainsClass(jarPath string, className string) bool {
	classFile := strings.ReplaceAll(className, ".", "/") + ".class"

	r, err := zip.OpenReader(jarPath)
	if err != nil {
		return false
	}
	defer r.Close()

	for _, file := range r.File {
		if file.Name == classFile {
			return true
		}
	}
	return false
}
